hydroflow/src/builder/surface/mod.rs

```diff
@@ -0,0 +1,159 @@
+//! Structs used to create the user-facing Surface API.
+//!
+//! Main user-facing traits are [`BaseSurface`], [`PullSurface`], and
+//! [`PushSurface`], which provide an iterator-like API with easy method
+//! chaining. The traits need to be imported when using the Surface API.
+//! **TODO: PRELUDE**
+//!
+//! * [`BaseSurface`] provides linear chaining methods like [`BaseSurface::map`], [`BaseSurface::filter`], etc..
+//! * [`PullSurface`] provides methods to combine multiple input streams: [`PullSurface::chain`], [`PullSurface::join`].
+//!     * To switch to push, call [`PullSurface::pivot`].
+//! * [`PushSurface`] provides sink chaining methods and methods to split into multiple output streams: [`PushSurface::tee`], [`PushSurface::for_each`].
+//!
+//! For implementation info see [super].
+
+use super::build::{PullBuild, PushBuild};
+use super::connect::{PullConnect, PushConnect};
+
+pub mod filter;
+pub mod flat_map;
+pub mod map;
+pub mod pivot;
+
+pub mod pull_chain;
+pub mod pull_handoff;
+pub mod pull_join;
+
+pub mod push_for_each;
+pub mod push_handoff;
+pub mod push_pivot;
+pub mod push_start;
+pub mod push_tee;
+
+use std::hash::Hash;
+
+use crate::scheduled::handoff::{HandoffList, HandoffListSplit};
+use crate::scheduled::type_list::Extend;
+
+/// Common trait shared between push and pull surface APIs.
+///
+/// Provides non-push/pull-specific chaining methods.
+pub trait BaseSurface {
```

Comment:
Nice that these are shared, I thought we'd have to duplicate these method definitions.